network/validator_tracker.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+const (
+	minCountEstimate               = 128
+	targetFalsePositiveProbability = .001
+	maxFalsePositiveProbability    = .01
+)
+
+var _ validators.SetCallbackListener = (*ValidatorTracker)(nil)
+
+func NewValidatorTracker(log logging.Logger) (*ValidatorTracker, error) {
+	tracker := &ValidatorTracker{
+		log:                        log,
+		connected:                  make(map[ids.NodeID]*ips.ClaimedIPPort),
+		connectedValidatorIndicies: make(map[ids.NodeID]int),
+	}
+	return tracker, tracker.resetBloom()
+}
+
+type ValidatorTracker struct {
+	log logging.Logger
+
+	lock                             sync.RWMutex
+	connected                        map[ids.NodeID]*ips.ClaimedIPPort
+	validators                       set.Set[ids.NodeID]
+	connectedValidatorIndicies       map[ids.NodeID]int
+	connectedValidators              []*ips.ClaimedIPPort
+	connectedValidatorsBloom         *bloom.Filter       // Bloom we use to pull new connections from other peers
+	connectedValidatorsAdditions     set.Set[ids.NodeID] // Set of validatorIDs that have currently been added to the bloom filter
+	connectedValidatorsBloomSalt     ids.ID
+	maxConnectedValidatorsBloomCount int
+}
+
+func (v *ValidatorTracker) Connected(nodeID ids.NodeID, ip *ips.ClaimedIPPort) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	v.connected[nodeID] = ip
+	if v.validators.Contains(nodeID) {
+		v.addConnectedValidator(nodeID, ip)
+	}
+}
+
+func (v *ValidatorTracker) Disconnected(nodeID ids.NodeID) *ips.ClaimedIPPort {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	ip, ok := v.connected[nodeID]
+	if !ok {
+		return nil
+	}
+
+	delete(v.connected, nodeID)
+	v.removeConnectedValidator(nodeID)
+	return ip
+}
+
+func (v *ValidatorTracker) OnValidatorAdded(nodeID ids.NodeID, _ *bls.PublicKey, _ ids.ID, _ uint64) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	v.validators.Add(nodeID)
+	if ip, connected := v.connected[nodeID]; connected {
+		v.addConnectedValidator(nodeID, ip)
+	}
+}
+
+func (*ValidatorTracker) OnValidatorWeightChanged(ids.NodeID, uint64, uint64) {}
+
+func (v *ValidatorTracker) OnValidatorRemoved(nodeID ids.NodeID, _ uint64) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	v.validators.Remove(nodeID)
+	v.removeConnectedValidator(nodeID)
+}
+
+func (v *ValidatorTracker) addConnectedValidator(nodeID ids.NodeID, ip *ips.ClaimedIPPort) {
+	v.connectedValidatorIndicies[nodeID] = len(v.connectedValidators)
+	v.connectedValidators = append(v.connectedValidators, ip)
+
+	// If we have already added this node to the bloom filter, wait until
+	// manually resetting the bloom filter to update the entry.
+	if v.connectedValidatorsAdditions.Contains(nodeID) {
+		return
+	}
+
+	// If the validator set is growing rapidly, we should increase the size of
+	// the bloom filter.
+	if count := v.connectedValidatorsBloom.Count(); count >= v.maxConnectedValidatorsBloomCount {
```

Comment:
This was one of the cases I wanted to ensure was handled (rapid increase in connected peers around startup) 👍